_posts/2023-04-19_esextractor.md

```diff
@@ -0,0 +1,95 @@
+
+## What is ESExtractor ?
+
+ESExtractor aims to be a simple elementary stream extractor. For the first revision it was able to extract video data from
+a file in the [NAL standard](https://en.wikipedia.org/wiki/Network_Abstraction_Layer).
+The first official release was [0.2.4](https://github.com/Igalia/ESExtractor/releases/tag/release-v0.2.4). In this release,
+only the NAL was supported with both H264 and H265 streams supported.
+
+As Vulkan Video aims to support more than H264 and H265 including format such as AV1 or VP9, the ESExtractor had to evoluate
+and support multiple format. A redesign has been initiated to support multiple format and is now available in the version
+0.3.0.
+
+## How ESExtractor works
+
+A simple C interface has been chosen to interface the extractor allowing to use the project in any other projects/languages etc.
+
+### es_extractor_new
+
+```
+ESExtractor extractor = es_extractor_new(filePath, "options"));
+```
+
+This interface returns the main object which will give you access to the packets according to the file path and the options given in arguments.
+This interface returns a ready to use object where the stream has been initially parsed to determine the kind of video during
+the object creation.
+
+Then you can check the video format with:
+
+```
+ESEVideoFormat eVideoFormat = es_extractor_video_format(extractor);
+```
+
+or the video codec with:
+
+```
+ESEVideoCodec eVideoCodec = es_extractor_video_codec(extractor);
+```
+
+It supports H264, H265, AV1 and VP9 for now.
+
+### es_extractor_read_packet
+
+This API is the main function to retrieve the available packets from the file. Each time this API is called,
+the library will return the next available packet according to the format and the specific alignment (ie NAL) and
+a status to let the application decides what to do next. The packet should be freed using `es_extractor_clear_packet`.
+
+
+## Choose C++ language as a good compromised
+
+As C++ is offering a lot of facilities and was easy to interface with the CTS.
+We have been using the facilites such as iostream and vectors to navigate through the stream which allows stability and a very low dependency
+to the system arch and any idea of addditional libraries.
+
+
+## Create a complete CI solution with github
+
+To secure the library usage, we have implementing a testing framework in addition to a CI infrastructure
+As github offers a very powerfull worklow, we decided to use this platform to test the library on various arch and platform.
+The [CI](https://github.com/Igalia/ESExtractor/actions) is now configured to release package for 64 and 32 bits on Linux and Windows.
+
+
+## Challenges
+
+To keep it simple and without any dependencies, the challenge was to a language
+which is quite well supported, fast and compatible with the CTS.
+To be fully integrated with CTS, we had to keep the API very simple and the source code, well partitioned.
```

Comment:
and dependency-free, the challenge was to choose a language which is well supported, ...

and the source code well partitioned